### DISCUSSION TITLE
Svelte 5 : Compose Builders [Tabs + Button]

### DISCUSSION THREAD

**Discussion by Stephane-Pires** [melt-ui/melt-ui#1275] [Q&A] [answered]
Hi. 

Thanks you for this wonderful library. Viewing component has builder with behavior has been really helpful to me in my current project. Especially regarding the accessibility. Overall I found the builder really well made. 

I recently encountered a potential  issue/incompatibility, regarding MeltUI builders and Svelte5

## My Goal ⭐ 

I try to compose `createTabs` with a `<Button />` I have made. 

## TLDR : 

- How do you usually compose `builders` , especially in the case of `createTabs ` + `trigger` ( I have already read : [https://melt-ui.com/docs/usage](usage))
-  Are builders expected to work in Svelte 5 ?

### In Svelte 4 🏃 (working so far)

```svelte
// Tabs.svelte component (instantiate createTabs)

const {
		elements: { root, list, content, trigger },
		states: { value: selectedTabId }
	} = createTabs({
		defaultValue: tabs.find((tab) => tab.id === tabId)!.id,
		onValueChange
	});

// ... some code

{#each tabs as tabItem}
			<div class={tabsStyles({ state: tabState($selectedTabId === tabItem.id, isLocked) }).tab()}>
				<Button
					builder={{
						kind: BUTTON_BUILDER_KIND.FUNCTION,
						element: trigger,
						parameter: tabItem.id
					}}
                                       // ... other props
                                />
{/each}


// Button.svelte component

export let builder:
		| { kind: Extract<ButtonBuilderKind, "primitive">; element: AnyMeltElement }
		| { kind: Extract<ButtonBuilderKind, "function">; element: AnyMeltElement; parameter: any } = {
		kind: BUTTON_BUILDER_KIND["PRIMITIVE"],
		element: emptyMeltElement
	};

	const { kind, element } = builder;

<button
	use:melt={kind === BUTTON_BUILDER_KIND.PRIMITIVE ? $element : $element(builder.parameter)}
        //...  otherProps.
</button>
```

This way of composing is **WORKING** in Svelte 4

_I'm really open minded about better way to compose it, because it feel weird, and not natural to me, to do _

```ts
export let builder:
		| { kind: Extract<ButtonBuilderKind, "primitive">; element: AnyMeltElement }
		| { kind: Extract<ButtonBuilderKind, "function">; element: AnyMeltElement; parameter: any } = {
		kind: BUTTON_BUILDER_KIND["PRIMITIVE"],
		element: emptyMeltElement
	};
```

### Svelte 5 🔴  (not working so far)

```svelte
// +Layout.svelte component (instantiate createTabs)

	const {
		elements: { root, list, content, trigger },
		states: { value }
	} = createTabs();

// ... some code
{#each NAV_ITEM as { id, icon, label, route, isActive }}
			<Button
				builder={{
					kind: BUTTON_BUILDER_KIND.FUNCTION,
					element: trigger,
					parameter: id
				}}
                               // other props
                              />
{/each}


// Button.svelte component

export let builder:
		| { kind: Extract<ButtonBuilderKind, "primitive">; element: AnyMeltElement }
		| { kind: Extract<ButtonBuilderKind, "function">; element: AnyMeltElement; parameter: any } = {
		kind: BUTTON_BUILDER_KIND["PRIMITIVE"],
		element: emptyMeltElement
	};

	let {
               // otherProps,
		builder = {
			kind: BUTTON_BUILDER_KIND['PRIMITIVE'],
			element: emptyMeltElement
		},
	}: Props = $props();

<button
	use:melt={kind === BUTTON_BUILDER_KIND.PRIMITIVE ? $element : $element(builder.parameter)}
        //...  otherProps.
</button>
```

This is **NOT WORKING**

<img width="1263" alt="Capture d’écran 2024-10-18 à 09 11 19" src="https://github.com/user-attachments/assets/b76494d0-bfd5-4e32-801a-ab5d0abc1de2">

Thanks for taking the time to read me 👍 🙏

**Comment by TGlide** [comment, marked as answer]
We're working on a new version of Melt for Svelte 5. There are some issues with the current version, unfortunately! 

A public beta will be out soon-ish

**Reply by Stephane-Pires** [reply]
Thanks for your answer. 

> We're working on a new version of Melt for Svelte 5.

This solve my question,  can't wait to see what you have done 👍

**Reply by Stephane-Pires** [reply]
Update : Latest news 

https://www.thomasglopes.com/articles/melt-ui-runes-is-open-source

**Reply by esafak** [reply]
Is there a way we can keep track of progress?

**Reply by TGlide** [reply]
Yup, on Discord or follow the new repo
